### ISSUE TITLE
Dynamic sampling uses exact float zero comparison instead of epsilon guard — near-zero variance groups not filtered

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/autonomous-dev#415]
## Problem

`_filter_dynamic_sampling` in `src/realign/backends/pytorch/grpo_trainer.py` line 295:

```python
if group_std.item() == 0.0:
    keep_mask[mask] = False
```

Comparing a float32 standard deviation to exactly `0.0` is fragile. If all rewards in a group are intended to be identical but differ at floating-point precision (e.g., `[1.0, 1.0, 1.0 + 1e-9]` due to reward model output rounding, log-prob accumulation, or float32 arithmetic), `std > 0.0` and the group is not filtered. These groups provide negligible learning signal and contribute to entropy collapse — exactly what dynamic sampling is meant to prevent.

## Rule Violated

Global CLAUDE.md: precision in algorithmic implementation should match documented intent. The DAPO paper describes filtering groups where reward variance is negligible, not strictly zero.

## Evidence

- `grpo_trainer.py` line 295: `if group_std.item() == 0.0`
- `TestDynamicSampling` in `test_grpo_dapo.py` only tests exact-zero variance (`[1.0, 1.0, 1.0]`), no near-zero variance test
- Reward models commonly output values like `0.99999994` and `1.0` for identical responses due to float32 precision

## Suggested Fix

Replace the exact comparison with an epsilon guard:

```python
if group_std.item() < 1e-8:
    keep_mask[mask] = False
```

Add a test:

```python
def test_dynamic_sampling_filters_near_zero_variance(self, dapo_config):
    """Groups with near-zero (but nonzero) variance should also be filtered."""
    trainer = _create_trainer(dapo_config)
    # Near-zero variance — float32 arithmetic on identical values
    rewards = torch.tensor([1.0, 1.0, 1.0 + 1e-9])
    group_indices = torch.tensor([0, 0, 0])
    keep_mask = trainer._filter_dynamic_sampling(rewards, group_indices)
    assert not keep_mask.any()
```

---
*Filed automatically by continuous-improvement-analyst*
